src/signals/incident/services/map-controls-to-params/index.js

```diff
@@ -22,36 +36,61 @@ const mapControlsToParams = (incident, wizard) => {
     params.incident_date_start = date.format();
   }
 
+  const category_url = incident && incident.subcategory_link ? new URL(incident.subcategory_link).pathname : '';
   const map = [];
   let mapMerge = {};
   forEach(wizard, (step) => {
-    const controls = step.form && step.form.controls;
+    let controls = {};
+    if (step.formFactory && isFunction(step.formFactory)) {
+      const form = step.formFactory(incident);
+      controls = form && form.controls;
+    } else {
+      controls = step.form && step.form.controls;
+    }
     forEach(controls, (control, name) => {
-      if (control.meta && control.meta.path) {
+      const value = incident[name];
+      const meta = control.meta;
+
+      if (meta && meta.path) {
         map.push({
-          path: control.meta.path,
-          value: incident[name]
+          path: meta.path,
+          value
         });
       }
 
-      if (control.meta && control.meta.pathMerge && incident[name]) {
-        mapMerge = {
-          ...mapMerge,
-          [control.meta.pathMerge]: {
-            ...mapMerge[control.meta.pathMerge],
-            [control.meta.label || name]: incident[name]
-          }
-        };
+      if (meta && meta.isVisible && meta.pathMerge) {
+        const answer = setValue(value);
+        if (answer || answer === 0) {
+          mapMerge = {
+            ...mapMerge,
+            [meta.pathMerge]: [
+              ...(mapMerge[meta.pathMerge] || []),
+              {
+                id: name,
+                label: meta.label,
+                category_url,
+                answer
+              }
+            ]
+          };
+        }
       }
     });
   });
 
   forEach(map, (item) => {
-    set(params, item.path, item.value);
+    let itemValue = setValue(item.value);
+    if (itemValue || itemValue === 0) {
+      if (isObject(itemValue) && itemValue.id) {
+        itemValue = itemValue.id;
+      }
+
+      set(params, item.path, itemValue);
+    }
   });
 
-  forEach(mapMerge, (item, key) => {
-    set(params, key, item);
+  forEach(mapMerge, (value, key) => {
+    set(params, key, value);
```

Comment:
I understand the need. But overall it's getting very very complicated. Possibly we want to split this file up into multiple smaller functions. That however is simply mitigating the underlying problem that the form logic is getting to complex.

Reply:
I totally agree. I will do this afternoon